### PR TITLE
unsymlink-lib: do not prepend vardb paths with EPREFIX.

### PR DESCRIPTION
**Pull request by heroxbd**
The file paths in vardb have EPREFIX.  Prepending them again results
in double prefix.

EROOT splitted into ROOT and EPREFIX, vardb paths are prepended with
ROOT.

Bug: https://bugs.gentoo.org/753743
Signed-off-by: Benda Xu <heroxbd@gentoo.org>

### CODE REVIEW

File: unsymlink-lib [hunk 1]
```diff
@@ -193,7 +193,8 @@ class MigrationState(object):
                 # files contained within them
                 if details[0] == 'dir':
                     continue
-                f = os.path.join(self.eroot, f.encode(_encodings['fs']).lstrip(b'/'))
+                f = f.encode(_encodings['fs'])
```
Comment (mgorny) [inline unsymlink-lib]: I think we need to split `self.eroot` into `ROOT` and `EPREFIX` here, and still prepend `ROOT` but I guess this is fine for the time being.

Comment (heroxbd) [reply unsymlink-lib]: Got an idea: read EPREFIX from portage.const and reconstruct ROOT. Let me try again.

Comment (heroxbd) [reply unsymlink-lib]: Made a split of eroot into root and eprefix.  What do you think?

File: unsymlink-lib [hunk 2]
```diff
@@ -164,7 +164,11 @@ class MigrationState(object):
         self.eroot = eroot
 
     def analyze(self, usr_merge, real_prefixes):
-        from portage import create_trees, _encodings
+        from portage import create_trees, const, _encodings
+
+        eprefix = const.EPREFIX.encode(_encodings['fs'])
```
Comment (mgorny) [inline unsymlink-lib]: I don't think this is entirely safe. I would have preferred if you moved `create_trees()` earlier, and grabbed it from the config instance.

File: unsymlink-lib [hunk 3]
```diff
@@ -193,7 +197,8 @@ class MigrationState(object):
                 # files contained within them
                 if details[0] == 'dir':
                     continue
-                f = os.path.join(self.eroot, f.encode(_encodings['fs']).lstrip(b'/'))
+                f = f.encode(_encodings['fs'])
```
Comment (mgorny) [inline unsymlink-lib]: Are you really, really sure we don't need stripping?

File: unsymlink-lib [hunk 4]
```diff
@@ -164,7 +164,11 @@ class MigrationState(object):
         self.eroot = eroot
 
     def analyze(self, usr_merge, real_prefixes):
-        from portage import create_trees, _encodings
+        from portage import create_trees, const, _encodings
+
+        eprefix = const.EPREFIX.encode(_encodings['fs'])
+        assert self.eroot.endswith(eprefix), "EROOT must be ROOT/EPREFIX."
+        root = self.eroot[:-len(eprefix)]
```
Comment (mgorny) [inline unsymlink-lib]: Not sure if we don't have to account for extra slashes, esp. using `--root /foo/bar/`.

File: unsymlink-lib (hunk 3 above)
Comment (heroxbd) [reply unsymlink-lib]: Ah, I should put back stripping.

File: unsymlink-lib (hunk 4 above)
Comment (heroxbd) [reply unsymlink-lib]: Let's do an `os.path.normpath()` first.

File: unsymlink-lib (hunk 2 above)
Comment (heroxbd) [reply unsymlink-lib]: Agreed, EPREFIX could change with environment but portage.const.EPREFIX is always a constant.  They are not equal.

**Comment by mgorny** [comment]
Are you done making your changes and testing?

**Comment by heroxbd** [comment]
> Are you done making your changes and testing?

Yes, this is the final version.

**Comment by heroxbd** [comment]
Thanks Michał.
